Windows: The command line is (still) too long

**Describe the bug**

Pull request #2035 resolved the Windows error "The command line is too long" when a JBang Kotlin script is run from PowerShell or CMD. However the issue is still unresolved when the script is run from Bash on Windows.


**To Reproduce**

*Test01.sh* 

```bash
./build/install/jbang/bin/jbang version
./build/install/jbang/bin/jbang cache clear
./build/install/jbang/bin/jbang ./Test01.kt
```

```
wfouche@DESKTOP-IOB82O4 MINGW64 ~/IdeaProjects/jbang (main)
$ ./Test01.sh
0.126.1.12
[jbang] Clearing cache for urls
[jbang] Clearing cache for jars
[jbang] Clearing cache for kotlincs
[jbang] Clearing cache for groovycs
[jbang] Clearing cache for scripts
[jbang] Clearing cache for stdins
[jbang] Clearing cache for deps
[jbang] Downloading Kotlin 2.0.21. Be patient, this can take several minutes...
[jbang] Installing Kotlin 2.0.21...
[jbang] Resolving dependencies...
[jbang]    org.springframework.boot:spring-boot-starter-web:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-test:3.4.6
[jbang]    org.springframework.boot:spring-boot-configuration-processor:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-activemq:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-actuator:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-amqp:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-aop:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-artemis:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-batch:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-cache:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-cassandra:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-cassandra-reactive:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-couchbase:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-couchbase-reactive:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-elasticsearch:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-jdbc:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-jpa:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-ldap:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-mongodb:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-mongodb-reactive:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-neo4j:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-r2dbc:3.4.6
[jbang]    org.jetbrains.kotlin:kotlin-stdlib:2.0.21
[jbang] Dependencies resolved
[jbang] Building jar for Test01.kt...
The command line is too long.
[jbang] [ERROR] Error during compile
```

**JBang version**

Main branch, 0.126.12

**Additional information**

*Test01.kt*

```Kotlin
///usr/bin/env jbang "$0" "$@" ; exit $?

//DEPS org.springframework.boot:spring-boot-starter-web:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-test:3.4.6
//DEPS org.springframework.boot:spring-boot-configuration-processor:3.4.6
//DEPS org.springframework.boot:spring-boot-starter:3.4.6
//DEPS org.springframework.boot:spring-boot-starter:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-activemq:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-actuator:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-amqp:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-aop:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-artemis:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-batch:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-cache:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-data-cassandra:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-data-cassandra-reactive:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-data-couchbase:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-data-couchbase-reactive:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-data-elasticsearch:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-data-jdbc:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-data-jpa:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-data-ldap:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-data-mongodb:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-data-mongodb-reactive:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-data-neo4j:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-data-r2dbc:3.4.6

fun main(args: Array<String>) {
    println("Kotlin: Hello World");
}
```

*Test01.cmd*

```
call .\build\install\jbang\bin\jbang version
call .\build\install\jbang\bin\jbang cache clear
call .\build\install\jbang\bin\jbang .\Test01.kt
```

*Test01.cmd - output*

```
0.126.1.12
[jbang] Clearing cache for urls
[jbang] Clearing cache for jars
[jbang] Clearing cache for kotlincs
[jbang] Clearing cache for groovycs
[jbang] Clearing cache for scripts
[jbang] Clearing cache for stdins
[jbang] Clearing cache for deps
[jbang] Downloading Kotlin 2.0.21. Be patient, this can take several minutes...
[jbang] Installing Kotlin 2.0.21...
[jbang] Resolving dependencies...
[jbang]    org.springframework.boot:spring-boot-starter-web:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-test:3.4.6
[jbang]    org.springframework.boot:spring-boot-configuration-processor:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-activemq:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-actuator:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-amqp:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-aop:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-artemis:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-batch:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-cache:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-cassandra:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-cassandra-reactive:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-couchbase:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-couchbase-reactive:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-elasticsearch:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-jdbc:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-jpa:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-ldap:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-mongodb:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-mongodb-reactive:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-neo4j:3.4.6
[jbang]    org.springframework.boot:spring-boot-starter-data-r2dbc:3.4.6
[jbang]    org.jetbrains.kotlin:kotlin-stdlib:2.0.21
[jbang] Dependencies resolved
[jbang] Building jar for Test01.kt...
Kotlin: Hello World
```

--- a/src/main/java/dev/jbang/source/buildsteps/CompileBuildStep.java
+++ b/src/main/java/dev/jbang/source/buildsteps/CompileBuildStep.java
@@ -118,7 +118,7 @@ public abstract class CompileBuildStep implements Builder<Project> {
 
 	protected void runCompiler(List<String> optionList) throws IOException {
 		runCompiler(CommandBuffer.of(optionList)
-			.applyWindowsMaxLengthLimit()
+			.applyWindowsMaxProcessLimit()
 			.asProcessBuilder()
 			.inheritIO());
 	}

--- a/src/main/java/dev/jbang/source/buildsteps/NativeBuildStep.java
+++ b/src/main/java/dev/jbang/source/buildsteps/NativeBuildStep.java
@@ -72,7 +72,7 @@ public class NativeBuildStep implements Builder<Project> {
 		Util.verboseMsg("native-image: " + String.join(" ", optionList));
 
 		ProcessBuilder pb = CommandBuffer.of(optionList)
-			.applyWindowsMaxLengthLimit()
+			.applyWindowsMaxProcessLimit()
 			.asProcessBuilder()
 			.inheritIO();
 

--- a/src/main/java/dev/jbang/source/generators/JarCmdGenerator.java
+++ b/src/main/java/dev/jbang/source/generators/JarCmdGenerator.java
@@ -200,7 +200,7 @@ public class JarCmdGenerator extends BaseCmdGenerator<JarCmdGenerator> {
 	protected String generateCommandLineString(List<String> fullArgs) throws IOException {
 		return CommandBuffer.of(fullArgs)
 			.shell(shell)
-			.applyWindowsMaxLengthLimit(CommandBuffer.MAX_LENGTH_WINCLI)
+			.applyWindowsMaxCliLimit()
 			.asCommandLine();
 	}
 

--- a/src/main/java/dev/jbang/util/CommandBuffer.java
+++ b/src/main/java/dev/jbang/util/CommandBuffer.java
@@ -84,15 +84,41 @@ public class CommandBuffer {
 		return CommandBuffer.of(arguments.get(0), "@" + argsFile);
 	}
 
-	public CommandBuffer applyWindowsMaxLengthLimit() throws IOException {
-		int maxLength = MAX_LENGTH_WINPROCBUILDER;
-		if (Util.isWindows()) {
-			maxLength = MAX_LENGTH_WINCLI;
+	/**
+	 * Determines if it's necessary to switch to an args file depending on the OS
+	 * and Shell being used and the command to run.
+	 * 
+	 * @return either this <code>CommandBuffer</code> or a new one using an args
+	 *         file.
+	 * @throws IOException throws an exception if the args file could not be
+	 *                     created.
+	 */
+	public CommandBuffer applyWindowsMaxProcessLimit() throws IOException {
+		String cmd = arguments.get(0).toLowerCase();
+		if (cmd.endsWith(".bat") || cmd.endsWith(".cmd")) {
+			return applyWindowsMaxCliLimit();
 		}
-		return applyWindowsMaxLengthLimit(maxLength);
+		return applyWindowsMaxLengthLimit(MAX_LENGTH_WINPROCBUILDER);
 	}
 
-	public CommandBuffer applyWindowsMaxLengthLimit(int maxLength) throws IOException {
+	/**
+	 * Determines if it's necessary to switch to an args file depending on the OS
+	 * being used. Use this when its known beforehand that on Windows the command
+	 * can only use the more limited command size available to the command line
+	 * (about 8Kb). This is normally the case when trying to run batch files for
+	 * example. But in the case of JBang it's also used when returning the command
+	 * that should be executed by the startup script.
+	 * 
+	 * @return either this <code>CommandBuffer</code> or a new one using an args
+	 *         file.
+	 * @throws IOException throws an exception if the args file could not be
+	 *                     created.
+	 */
+	public CommandBuffer applyWindowsMaxCliLimit() throws IOException {
+		return applyWindowsMaxLengthLimit(MAX_LENGTH_WINCLI);
+	}
+
+	private CommandBuffer applyWindowsMaxLengthLimit(int maxLength) throws IOException {
 		String args = asCommandLine();
 		// Check if we can and need to use @-files on Windows
 		if (args.length() > maxLength && Util.isWindows()) {

--- a/src/test/java/dev/jbang/util/TestCommandBuffer.java
+++ b/src/test/java/dev/jbang/util/TestCommandBuffer.java
@@ -7,6 +7,7 @@ import static org.hamcrest.Matchers.*;
 import java.io.IOException;
 
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.condition.DisabledOnOs;
 import org.junit.jupiter.api.condition.EnabledOnOs;
 import org.junit.jupiter.api.condition.OS;
 
@@ -46,10 +47,43 @@ public class TestCommandBuffer extends BaseTest {
 
 	@Test
 	@EnabledOnOs(OS.WINDOWS)
-	void testApplyWindowsMaxLengthLimitExe() throws IOException {
-		ProcessBuilder pb = CommandBuffer.of(argsTooLong("foo.exe"))
+	void testApplyWindowsMaxProcessLimitExe1() throws IOException {
+		ProcessBuilder pb = CommandBuffer.of(argsTooLong("foo.exe", 1000))
 			.shell(Util.Shell.cmd)
-			.applyWindowsMaxLengthLimit()
+			.applyWindowsMaxProcessLimit()
+			.asProcessBuilder();
+		assertThat(pb.command().size(), greaterThan(2));
+		assertThat(pb.command().get(1), not(anyOf(startsWith("@"), startsWith("\"@"))));
+	}
+
+	@Test
+	@EnabledOnOs(OS.WINDOWS)
+	void testApplyWindowsMaxProcessLimitExe2() throws IOException {
+		ProcessBuilder pb = CommandBuffer.of(argsTooLong("foo.exe", 4000))
+			.shell(Util.Shell.cmd)
+			.applyWindowsMaxProcessLimit()
+			.asProcessBuilder();
+		assertThat(pb.command().size(), equalTo(2));
+		assertThat(pb.command().get(1), anyOf(startsWith("@"), startsWith("\"@")));
+	}
+
+	@Test
+	@DisabledOnOs(OS.WINDOWS)
+	void testApplyOthersMaxProcessLimit() throws IOException {
+		ProcessBuilder pb = CommandBuffer.of(argsTooLong("foo", 4000))
+			.shell(Util.Shell.bash)
+			.applyWindowsMaxProcessLimit()
+			.asProcessBuilder();
+		assertThat(pb.command().size(), greaterThan(2));
+		assertThat(pb.command().get(1), not(anyOf(startsWith("@"), startsWith("\"@"))));
+	}
+
+	@Test
+	@EnabledOnOs(OS.WINDOWS)
+	void testApplyWindowsMaxProcessLimitBat() throws IOException {
+		ProcessBuilder pb = CommandBuffer.of(argsTooLong("foo.bat", 1000))
+			.shell(Util.Shell.cmd)
+			.applyWindowsMaxProcessLimit()
 			.asProcessBuilder();
 		assertThat(pb.command().size(), equalTo(2));
 		assertThat(pb.command().get(1), anyOf(startsWith("@"), startsWith("\"@")));
@@ -57,10 +91,10 @@ public class TestCommandBuffer extends BaseTest {
 
 	@Test
 	@EnabledOnOs(OS.WINDOWS)
-	void testApplyWindowsMaxLengthLimitBat() throws IOException {
-		ProcessBuilder pb = CommandBuffer.of(argsTooLong("foo.bat"))
-			.shell(Util.Shell.cmd)
-			.applyWindowsMaxLengthLimit()
+	void testApplyWindowsMaxProcessLimitBatFromBash() throws IOException {
+		ProcessBuilder pb = CommandBuffer.of(argsTooLong("foo.bat", 1000))
+			.shell(Util.Shell.bash)
+			.applyWindowsMaxProcessLimit()
 			.asProcessBuilder();
 		assertThat(pb.command().size(), equalTo(2));
 		assertThat(pb.command().get(1), anyOf(startsWith("@"), startsWith("\"@")));
@@ -68,10 +102,21 @@ public class TestCommandBuffer extends BaseTest {
 
 	@Test
 	@EnabledOnOs(OS.WINDOWS)
-	void testApplyWindowsMaxLengthLimitCmd() throws IOException {
-		ProcessBuilder pb = CommandBuffer.of(argsTooLong("foo.cmd"))
+	void testApplyWindowsMaxProcessLimitCmd() throws IOException {
+		ProcessBuilder pb = CommandBuffer.of(argsTooLong("foo.cmd", 1000))
 			.shell(Util.Shell.cmd)
-			.applyWindowsMaxLengthLimit()
+			.applyWindowsMaxProcessLimit()
+			.asProcessBuilder();
+		assertThat(pb.command().size(), equalTo(2));
+		assertThat(pb.command().get(1), anyOf(startsWith("@"), startsWith("\"@")));
+	}
+
+	@Test
+	@EnabledOnOs(OS.WINDOWS)
+	void testApplyWindowsMaxProcessLimitCmdFromBash() throws IOException {
+		ProcessBuilder pb = CommandBuffer.of(argsTooLong("foo.cmd", 1000))
+			.shell(Util.Shell.bash)
+			.applyWindowsMaxProcessLimit()
 			.asProcessBuilder();
 		assertThat(pb.command().size(), equalTo(2));
 		assertThat(pb.command().get(1), anyOf(startsWith("@"), startsWith("\"@")));
@@ -97,8 +142,8 @@ public class TestCommandBuffer extends BaseTest {
 		assertThat(cmd, is(cmd2));
 	}
 
-	private String[] argsTooLong(String cmd) {
-		String[] args = new String[1000];
+	private String[] argsTooLong(String cmd, int cnt) {
+		String[] args = new String[cnt];
 		args[0] = cmd;
 		for (int i = 1; i < args.length; i++) {
 			args[i] = "argument " + i;

